### ISSUE TITLE
Limit concurrent http_get requests

### ISSUE THREAD

**Issue by jsicherman** [gaborcsardi/async#59]
Hi,

I'm making multiple API queries in parallel via. something like

```
lapply(QUERIES, function(query) {
    http_get(query)$then({
        ....
        if(SOMETHING) {
            http_get(...)$then(...)
        } else {
            async_constant(...)
        }
    })
}) %>% { when_all(.list = .)$then(function(x) unlist(x, recursive = F)) }
```

In fact, it's even more complex as I have potentially nested `lapply`s and thus potentially making hundreds-thousands of asynchronous queries. I'm not too familiar with the networking side of things, but I'm concerned that in doing this, I can potentially overwhelm the server with requests all at once.

I'm wondering if
1. {async} has built-in limits on how many `http_get` can be pending at once
2. Is this actually a problem? If so is there, any solution? I imagine potentially `delay(...)$then(...)` every N loop iterations, but this seems like a dirty workaround.

Thanks

**Comment by gaborcsardi** [comment]
Good question. Yes, async uses the `multi_*` function from the curl package to do HTTP, and these have built-in limits, see `?curl::new_pool()`. They are currently maximum 100 connections in total, and maximum 6 connections to the same host. If both the client and the server support HTTP/2 then HTTP/2 multiplexing is used to use the same connection for multiple requests.

It would indeed make sense to expose these and other libcurl parameters.

**Comment by gaborcsardi** [comment]
And as for your second question, it is usually not, because web servers are used to clients making multiple connections at the same time, because web servers do that as well.

However, I can imagine that some web apis restrict the number of connections for a client, and then it may be a problem. So it is good to be able to change this.

**Comment by JZL** [comment]
I had the same issue but with the opposite problem. I was using async to do a batch of long-running API requests (all to the same host with different url paramaters) so was hitting the `curl::new_pool` host_con=6 limitation very heavily. I was submitting 50-60 requests at once but it was only putting 6 out at a time.

The curl limitations are pretty pernicious because even though I was browser/print debugging around the http_get call, everything was being submitted at once, it was curl which was rate limiting and I couldn't test that. It was lucky I was digging through the code, saw the curl pool calls, and thought to check its limitations but it took me a day of solid debugging to figure it out. 

I think exposing the curl paramaters to be changed would be very helpful, if only so that it shows up in documentation. Although personally I think 6 is too small so don't see why the default shouldn't be much higher. (I guess if scraping many different domains it could help to have per-domain limitations but that seems somewhat niche)

**Comment by gaborcsardi** [comment]
Closed by 4d063509e8cd4a1351bf9810d9631336ca483479

**Comment by gaborcsardi** [comment]
@JZL Let me know if this works for you now.
